test/controllers/sessions_controller_test.rb

```diff
@@ -0,0 +1,28 @@
+require "test_helper"
+
+class SessionsControllerTest < ActionDispatch::IntegrationTest
+
+  setup do 
+    OmniAuth.config.test_mode = true 
+    OmniAuth.config.mock_auth[:google] = OmniAuth::AuthHash.new({
+      provider:'google',
+      uid:'admin@email.com',
+      info: {
+        name:'admin',
+        email: '123@admin.com'
+        }
+      })
+  end
+    
+  test "should create session" do 
```

Comment:
The success condition in this test would be a redirect to `books_path` and a failure would be a redirect to `root_path`.

I also think there is a bug in `SessionsController.create()`. The error flash should be set in an `else` underneath `if user.username == 'admin' or user.medusa_admin?` and the outer check for `if auth_hash and auth_hash[:uid]` should be removed. I think.

Reply:
Thanks for pointing that out. 

Testing the success condition in this case would require testing the private/protected method for the `clear/return path`; is it generally okay to unit test protected methods?

Reply:
It's not possible to test `clear_and_return_return_path()` in a controller test (because there is no route to it) but it should be sufficient to just assume that it's going to redirect to `books_path` as long as there is no referer URL. Although you could test those conditions too.

Reply:
This is now updated to show a redirect to books_path upon successful sign in and a redirect to root upon failed sign in.